Moved password reset token to redis

niklabh opened the pull request:
Closes: https://github.com/paritytech/polkassembly/issues/599
Closes: https://github.com/paritytech/polkassembly/issues/631

Tbaut commented on auth-server/src/services/auth.ts:
```diff
@@ -557,36 +558,17 @@ export default class AuthService {
-			.first();
-
-		if (!resetToken) {
-			throw new AuthenticationError(messages.PASSWORD_RESET_TOKEN_NOT_FOUND);
```
This message should be removed because it's not used any more, except in the test that are actually failing.

niklabh commented:
Fixing tests moved to seperate issue:

https://github.com/paritytech/polkassembly/issues/632
https://github.com/paritytech/polkassembly/issues/631

Tbaut commented:
Setting the "on-ice" label. I really don't want to merge this PR before we have tests for it.
I'll work on #631

niklabh commented:
> Setting the "on-ice" label. I really don't want to merge this PR before we have tests for it.
> I'll work on #631

changed key to userId. Added tests.

Tbaut commented on auth-server/test/resolvers/mutation/requestResetPassword.spec.ts:
```diff
@@ -40,70 +36,15 @@ describe('requestResetPassword mutation', () => {
-	});
-
-	it('should not be able to change password with an invalid email', async () => {
+	it('should not be able to request reset password with an invalid email', async () => {
```
Let's move this test to resetPassword.

niklabh replied:
This file is testing requestResetPassword not resetPassword mutation. Shouldn't each mutation has sperate tests?

Tbaut replied:
That would be better, but as of now, things are all mixed, `resetPassword` has a test called "should be able to request a password reset with token set in redis" which should then rather be in `requetsResetPassword`.

niklabh replied:
done

Tbaut commented on auth-server/test/resolvers/mutation/resetPassword.spec.ts:
```diff
@@ -0,0 +1,114 @@
+		}
+	});
+
+	it('should not be able to reset password with a wrong token', async () => {
```
We should have this test before the one that works, with the valid token.

Tbaut commented on auth-server/test/resolvers/mutation/resetPassword.spec.ts:
```diff
@@ -0,0 +1,114 @@
+		}
+	});
+
+	it('should not be able to request reset password with an invalid email', async () => {
```
Let's have this test first.

Tbaut commented on auth-server/test/resolvers/mutation/resetPassword.spec.ts:
```diff
@@ -0,0 +1,114 @@
+	});
+
+	it('should not be able to change password with an expired token', async () => {
+		await redisDel(getPwdResetTokenKey(signupResult.user.id));
```
The token should have been deleted after the successful retrieval before, so there shouldn't be any need to remove the token here.
```suggestion
```

Tbaut commented on auth-server/test/resolvers/mutation/resetPassword.spec.ts:
```diff
@@ -0,0 +1,114 @@
+		}
+	});
+
+	it('should not be able to change password with an expired token', async () => {
```
See previous comment. We don't test the expiration here.
```suggestion
	it('should not be able to change password with token that was used already', async () => {
```

niklabh commented:
> I keep getting:
> 
> > 1. addressLogin mutation
> >    auth-test_1      |        should be able to login with addresss:
> >    auth-test_1      |      ForbiddenError: Session expired, please login again.
> 
> Also I've had another weird failure like this one (not reproducible every time.. looks like a local issue):
> 
> > auth-test_1      |   2) user query
> > auth-test_1      |        should return users with limit:
> > auth-test_1      |
> > auth-test_1      |       AssertionError: expected 607 to equal 593
> > auth-test_1      |       + expected - actual
> > auth-test_1      |
> > auth-test_1      |       -607
> > auth-test_1      |       +593
> > auth-test_1      |
> > auth-test_1      |       at /opt/test/test/resolvers/query/users.spec.ts:49:23
> > auth-test_1      |       at Array.forEach ()
> > auth-test_1      |       at Context. (test/resolvers/query/users.spec.ts:47:10)
> 
> Have you ever had this?

Cannot reproduce. Have merged master. Can you pull once and check.

Tbaut commented:
> Cannot reproduce. Have merged master. Can you pull once and check.

Yes, this was fixed in the lint PR. Testing again now.

Tbaut commented on front-end/src/screens/ResetPassword/index.tsx:
```diff
@@ -21,12 +21,14 @@ interface Props {
 const ResetPassword = ({ className }:Props): JSX.Element => {
 	const router = useRouter();
 	const { token } = router.query;
+	const userId = Number(new URLSearchParams(window.location.search).get('user_id'));
```
we should use the useRouter hook for both the token and the user_id.
ps: let's use `userId` rather than `user_id` if possible

niklabh replied:
done

Tbaut commented on auth-server/src/services/email.ts:
```diff
@@ -49,17 +48,14 @@ export const sendVerificationEmail = (user: User, token: EmailVerificationToken)
-		resetPasswordEmailTemplate,
-		{ resetUrl, username: user.name || '' }
-	);
+	const resetUrl = `${DOMAIN}/reset-password/${token}?user_id=${user.id}`;
```
let's use query string for both, `?token=abc&userId=123`

niklabh replied:
done
